torch/testing/_internal/common_methods_invocations.py

```diff
@@ -4,14 +4,133 @@
+        self.tests_to_skip = tests_to_skip
+
+    # Returns the op (currently only supports torch.<op_name>)
+    def getOp(self):
+        return getattr(torch, self.name)
+
+    # Returns the method variant of the op (torch.Tensor.<op_name>)
+    # or None if the op has no method variant
+    def getMethod(self):
+        return self.method_variant
+
+    # Returns the inplace variant of the op (torch.Tensor.<op_name>_)
+    # or None if the op has no inplace variant
+    def getInplace(self):
+        return self.inplace_variant
+
+    # Returns True if the test should be skipped and False otherwise
+    def shouldSkip(self, test_name, *, device, dtype):
+        if self.tests_to_skip is None:
+            return False
+
+        if device is None and dtype is None:
+            full_name = test_name
+        elif dtype is None:
+            assert device is not None
+            full_name = test_name + "_" + str(torch.device(device).type)
+        else:
+            assert device is not None
+            assert dtype is not None
+            full_name = test_name + "_" + str(torch.device(device).type) + "_" + str(dtype).split('.')[1]
+
+        for regex in self.tests_to_skip:
+            if re.match(regex, full_name):
+                return True
+
+        return False
+
+
+# Metadata class for unary "universal functions (ufuncs)" that accept a single
+# tensor and have common properties like:
+#   - they are elementwise functions
+#   - the input shape is the output shape
+#   - they typically have method and inplace variants
+#   - they typically support the out kwarg
+#   - they typically have NumPy or SciPy references
+# See NumPy's universal function documentation
+# (https://numpy.org/doc/1.18/reference/ufuncs.html) for more details
+# about the concept of ufuncs.
+class UnaryUfuncMeta(OpMeta):
+    def __init__(self,
+                 name,  # the string name of the function
+                 *,
+                 dtypes=get_common_float_types(),
+                 dtypesIfCPU=get_common_float_and_complex_types_plus_bfloat16(),
+                 dtypesIfCUDA=get_common_float_and_complex_types_plus_half(),
+                 dtypesIfROCM=get_all_float_types(),
+                 has_out_kwarg=True,
+                 **kwargs):
+        super(UnaryUfuncMeta, self).__init__(name,
+                                             dtypes=dtypes,
+                                             dtypesIfCPU=dtypesIfCPU,
+                                             dtypesIfCUDA=dtypesIfCUDA,
+                                             dtypesIfROCM=dtypesIfROCM,
+                                             has_out_kwarg=has_out_kwarg,
+                                             **kwargs)
+
+# Operator database
+op_db = [
+    UnaryUfuncMeta('cos',
+                   dtypesIfCUDA=get_all_float_and_complex_types()),
+    UnaryUfuncMeta('cosh',
+                   dtypesIfCPU=get_common_float_and_complex_types()),
```

Comment:
The fact that you have to do a ton of overwriting here makes me wonder if you shouldn't do this API via an additive/subtractive mechanism, rather than a straight override. The problem with a direct override, is if you adjust the default dtype set on the top level ufunc, you STILL have to adjust all of the places where you overrode things explicitly to add the new default. With something additive or subtractive you could update the defaults and everything here would get updated as well.

Reply:
In other words, do we get anything from a hierarchy? If not, let's get rid of it!

Reply:
Hm... can you elaborate? Like if I want to test float32 and float64, and the default is float32, then if I specify the override I always get what I want independent of the default. If, however, my model is +float64, then if the default changes to, for example, float16, I would get float16 and float64, which is not what I want.

Reply:
You presented a different situation than what I was thinking about. Let me address it first, and then I'll describe the situation I was thinking about.

You suggest a situation where I change the default from float32 to float16. First off, this isn't an accurate description of how the defaulting works in your PR, as is! When `dtypesIfCPU` is `None`, then you default to use `dtypes`, which itself is defaulted to `get_common_float_types`, which is `torch.float32` and `torch.float64`. Which brings me to my main point: it seems incredibly unlikely to me that you would ever change `get_common_float_types` to only return a `torch.float16`! So the concern here seems like a total non sequitur. Perhaps you were remarking about a situation where `dtype=` is *locally* specified to be `float32`, and then `dtypesIfCPU` is additive. But here the argument doesn't seem to hold water either; if I change `dtype` for an operator from `float32` to `float16` (once again, this seems like an *incredibly* unusual thing to do--but for the sake of discussion, let's say you do it), but I happen to know that CPU should *still* test on float32, I should be reasonably expected to go and add float32 to the lists in CPU (if it hasn't been added already)--in particular, the change is local (I update `dtype=`, and now I see that `dtypesIfCPU=` needs to be updated too).

I was thinking about a different form of evolution, which seems far more plausible to me. Imagine we are globally adding support for a new dtype (float16, bfloat16 and complex{64,128} are great examples of this actually happening). At some point, we may get to the point where we have enough support that it makes sense to globally turn on the dtype for most of our tests, and simply blacklist the cases where things do not work. What we'd prefer to do is update some global location to turn on tests everywhere, and then grind through the test suite looking for cases where things don't work. Under your design here, this wouldn't work, because all of the places where overrides are happening, they will directly overwrite your global update, and so you'll default to *not* testing (as opposed to defaulting to testing). In the absence of test time considerations, I think that it is better to default to overtesting, because you *do not notice* when you undertest.

Reply:
I appreciate your scenario and actually agree with your logic, but...

- This would then be inconsistent with the analogous decorators, which override
- Designing a UX to support both additions and subtractions seem very finicky

To handle the case you describe I would prefer we just change the getters' notion of "common float types," for example. But if we do add a new dtype I think we'll still have to update every function's dispatch table to support it, so analogously we may have to update their dtype entries in this op database. That seems OK to me?

Reply:
> I would prefer we just change the getters' notion of "common float types," for example.

But this is exactly the point: this wouldn't work if someone didn't actually use the common float types getter. (Perhaps MKL is buggy on complex64 and it needed to be disabled.) Perhaps you are implying that it is wrong for a test to ever NOT use one of the getters. If so, let's make it impossible for people to not use getters!

> But if we do add a new dtype I think we'll still have to update every function's dispatch table to support it, so analogously we may have to update their dtype entries in this op database.

Not if we edit an `AT_FORALL` macro! But I do agree that if we have a Greg/Vitaly style database (in the purest form of the word, not just a bunch of Python defs) this is a much less pressing matter, because there is some other canonical source of information where you are actually getting this information from, and so one would imagine you aren't going to just straight up duplicate this information in the test suite. The design concept here seems to be, as faithfully as possible, replicate how we determine dtype support in the actual implementation. If so, we should explicitly talk about this concept, and make users abide by it!

Reply:
> But this is exactly the point: this wouldn't work if someone didn't actually use the common float types getter. (Perhaps MKL is buggy on complex64 and it needed to be disabled.) Perhaps you are implying that it is wrong for a test to ever NOT use one of the getters. If so, let's make it impossible for people to not use getters!

OK, let's do this now and make the getters correspond 1:1 with a subset of the dispatch macros for consistency. I think that will resolve this issue?